### ISSUE TITLE
Formatting the outputted selection in query mode

### ISSUE THREAD

**Issue by NightMachinery** [Genivia/ugrep#53]
I have been trying to get ugrep to output the path, the line number, and the column number of the selection(s) made in the interactive mode. I tried `--format='%u%f%s%n%s%k'`, and `--format-end='%u%f%s%n%s%k'`. The first makes ugrep never find anything (I don't understand why.), and the second doesn't do anything I can notice. My usecase here is to be able to jump from selections in interactive ugrep searches into my text editor (emacs). `--line-number`, `--column-number` and `-H` aren't useful for this task, because `--line-number` and `--column-number` pollute what is shown on the interactive search (who wants to see the line numbers and column numbers strewn over the search results?), and `-H` becomes a heading that is not included in the output (to stdout).

What I want is, e.g., when I exit from this selection:

![image](https://user-images.githubusercontent.com/36224762/88335242-3c3a1400-cd48-11ea-89f1-e7901f6e647a.png)

To get this output in stdout (`,` is the separator, the last field is the column number which is `-1` for non-matching lines, and LINE_NUMBER is the selection's line number.):
```
dev/notes.md,LINE_NUMBER,-1
todo/dev/aria2.md,LINE_NUMBER,3
```

**Comment by genivia-inc** [comment]
> I tried --format='%u%f%s%n%s%k', and --format-end='%u%f%s%n%s%k'. The first makes ugrep never find anything (I don't understand why.)

The problem is that the format has no terminating newline `\n` or `%~` so the output is concatenated. The lines in the UI are shown when the output produces a line.

**Comment by NightMachinary** [comment]
[**@genivia-inc**](https://github.com/genivia-inc) commented on [Jul 24, 2020, 1:36 AM GMT+4:30](https://github.com/Genivia/ugrep/issues/53#issuecomment-663232621 "2020-07-23T21:06:35Z - Replied by Github Reply Comments"):
> > I tried --format='%u%f%s%n%s%k', and --format-end='%u%f%s%n%s%k'. The first makes ugrep never find anything (I don't understand why.)
> 
> The problem is that the format has no terminating newline `\n` or `%~` so the output is concatenated. The lines in the UI are shown when the output produces a line.

Yes, this works (adding this tip to the manpage would be a nice touch), but it does not solve the problem of changing the format of the output, while not changing what the user sees in the interactive query mode.

**Comment by genivia-inc** [comment]
I think it is an overkill to add a key combination for the query UI to change the `--format`. The `--format` option is more useful to produce results in batch mode, e.g. to pipe the results to another application when that application expects a specific input format. That's why it exists, basically. There are many ways that is a useful feature to have, but not so much for the query UI. Also colors are lacking. So I don't see the point. Or am I missing something?

**Comment by NightMachinary** [comment]
[**@genivia-inc**](https://github.com/genivia-inc) commented on [Jul 24, 2020, 3:31 AM GMT+4:30](https://github.com/Genivia/ugrep/issues/53#issuecomment-663272617 "2020-07-23T23:01:13Z - Replied by Github Reply Comments"):
> I think it is an overkill to add a key combination for the query UI to change the `--format`. The `--format` option is more useful to produce results in batch mode, e.g. to pipe the results to another application when that application expects a specific input format. That's why it exists, basically. There are many ways that is a useful feature to have, but not so much for the query UI. Also colors are lacking. So I don't see the point. Or am I missing something?

I have failed to convey my meaning, let me try again.

I want to change the format not **during** the query mode (so no key combinations needed at all), as that is indeed not useful and the current system is much better. I want to change the format of selections *when ugrep exits and outputs to stdout*. So basically I just want the `--format` option to be applied when ugrep is exiting. This helps me use ugrep as an intermediate tool in my scripts.

Some example usages:
- interactively searching files and deleting selected lines
- interactively searching files and copying all the files where you had selected at least one line, to the clipboard (useful for when one wants to share some specific files with someone from among a crowded repository)
- a replacement for this (unmaintained) [tool](https://github.com/aykamko/tag), which makes it easy to immediately jump to a search match in your favorite editor.

What I want is more or less the same functionality you have in mind for the batch mode, but usable in the interactive mode, as well. I think this will greatly improve the value proposition of ugrep, as it will then not be just a grep replacement, but a basic terminal UI building block. This was probably the secret sauce that has made fzf so popular.

**Comment by genivia-inc** [comment]
@NightMachinary 
> I want to change the format not during the query mode (so no key combinations needed at all), as that is indeed not useful and the current system is much better. I want to change the format of selections when ugrep exits and outputs to stdout. So basically I just want the --format option to be applied when ugrep is exiting. This helps me use ugrep as an intermediate tool in my scripts.

I get it. That is an excellent suggestion.

This can be done with a minor tweak to the query UI and I'll take a look into planning this. Basically, the `--format` will be delayed until the output is produced. To produce the output will require one additional final search to be performed by the query UI to apply `--format`.

I wonder if this should also apply to the `--xml`, `--json`, `--csv` options? I think it should be done in the same way `--format` is going to be delayed to the final output, because browsing these formats in the query UI isn't as good anyway. These actually just set the `--format` strings to produce the specified output.

**Comment by NightMachinary** [comment]
Thanks! I am looking forward to it.

I think you're right that the change needs to propagate to the similar
options. I don't know if you're planning to change the default behavior or
to introduce a new CLI flag ala `--format-output`, but I think the old
behavior might be useful in some circumstances as well, so ideally there'll
be two different format flags, one for before and one for after. One way to
reduce the complexity of managing `--json` etc is to add their format
strings to the manpage, so the user will just copy the desired format
string into the appropriate format flag. They will also serve as good
examples of using the format strings.

**Comment by genivia-inc** [comment]
I will put the format forms for `--csv`, `--json`, and `--xml` in the README to document these, instead of the man page, because the man page should not get too long and cluttered.

#### `--csv`

option           | argument
---------------- | ------------------
`--format`       | `'%[,]$%H%N%K%B%V%~%u'`

#### `--json`

option           | argument
---------------- | ------------------
`--format-begin` | `'['`
`--format-open`  | `'%,%~  {%~    %[,%~    ]$%["file": ]H"matches": ['`
`--format`       | `'%,%~      { %[, ]$%["line": ]N%["column": ]K%["offset": ]B"match": %J }%u'`
`--format-close` | `'%~    ]%~  }'`
`--format-end`   | `'%~]%~'`

#### `--xml`

option           | argument
---------------- | ------------------
`--format-begin` | `'<grep>%~'`
`--format-open`  | `'  <file%[]$%[ name=]H>%~'`
`--format`       | `'    <match%[\"]$%[ line=\"]N%[ column=\"]K%[ offset=\"]B>%X</match>%~%u'`
`--format-close` | `'  </file>%~'`
`--format-end`   | `'</grep>%~'`

**Comment by genivia-inc** [comment]
@NightMachinary

There is one part of the discussion that I am not sure if we are on the same page.

You mentioned "outputted selection" in the title. Do you mean the selection mode that is entered with ENTER to select lines? Note that if the results are reformatted with `--format` than all the ENTERed selected lines will disappear and be replaced by the formatted output, whatever that is. If the selected output is selected as "All" then the formatted output would be OK. But if specific lines are selected then there is no clear method or guaranteed approach to produce the same in formatted form I think.

**Comment by NightMachinary** [comment]
[**@genivia-inc**](https://github.com/genivia-inc) commented on [Jul 25, 2020, 12:56 AM GMT+4:30](https://github.com/Genivia/ugrep/issues/53#issuecomment-663719677 "2020-07-24T20:26:08Z - Replied by Github Reply Comments"):
> [@NightMachinary](https://github.com/NightMachinary)
> 
> There is one part of the discussion that I am not sure if we are on the same page.
> 
> You mentioned "outputted selection" in the title. Do you mean the selection mode that is entered with ENTER to select lines? Note that if the results are reformatted with `--format` than all the ENTERed selected lines will disappear and be replaced by the formatted output, whatever that is. If the selected output is selected as "All" then the formatted output would be OK. But if specific lines are selected then there is no clear method or guaranteed approach to produce the same in formatted form I think.

Hmm ... I am not familiar with the internals, so I can only speak qualitatively. In the usages that I envision, I only have need of two pieces of information per selected line: The line number and the file it is located in. I think outputting these two shouldn't be hard? I agree that the general format strings are rather featureful and it is not trivial to apply them to the selections only. On the other hand, formatting all the matched files and outputting them, ignoring the selections, does not seem very useful to me.

Perhaps it's best to separate the two workflows: A `--selection-location` mode that outputs `path:line_number` for the selections instead of their raw text, and a `--format-output` flag that applies the formatting to all the matches and outputs them all. In this second mode, selections would be meaningless, so they should be disabled.

A new idea just popped into my head; Adding a `--print-final-query` that prints the final user query as the first line of output, will let the script run that query on a new invocation of ugrep in batch-mode with the desired format string. Thus there will be no need for the `--format-output` flag, and changes to ugrep can be minimized, while allowing greater user flexibility.

**Comment by genivia-inc** [comment]
There is a lot of useful info here. One more: we could add a key mapping (e.g. a designated Alt key) to switch between normal view and formatted view in the UI. This gives the best of both worlds at a fingertip. Once the search is satisfactory in normal view, the user can make selections in the formatted output by switching with an Alt key and the pressing ENTER.

**Comment by NightMachinary** [comment]
[**@genivia-inc**](https://github.com/genivia-inc) commented on [Jul 25, 2020, 2:49 AM GMT+4:30](https://github.com/Genivia/ugrep/issues/53#issuecomment-663756379 "2020-07-24T22:19:00Z - Replied by Github Reply Comments"):
> There is a lot of useful info here. One more: we could add a key mapping (e.g. a designated Alt key) to switch between normal view and formatted view in the UI. This gives the best of both worlds at a fingertip. Once the search is satisfactory in normal view, the user can make selections in the formatted output by switching with an Alt key and the pressing ENTER.

Any updates on this?

**Comment by genivia-inc** [comment]
No update. Low on the TODO list. I don't think there is any urgency. This cannot be simple layered on top of the current query engine. A new option should be introduced so that `--format` is preserved. Also, a lot more code is necessary to pull this off, because the file and line selections made in the UI have to be "translated" to make selection in the formatted output when the query UI exits, by running the search a second time and only selecting the formatted output that was selected. This is not trivial, because we need an engine that produces output only for a select list of line numbers (the user's selections).

**Comment by genivia-inc** [comment]
I've been thinking about the design of this feature for a while. It is not as straightforward as it may seem. It's much more complicated. To be able to output `--format` results while selecting non-formatted lines in the UI cannot be done without adding limits to ugrep that I think will not help. The selection in the UI is based on the current output of the search engine. The selected results are output as selected lines. But in fact, selecting "select all" does not store all results in memory. Rather, the search engine continues to produce the rest of the output without having to store everything in memory, which makes this efficient and also not memory bound. I think it is more of a luxury to have this feature, but not a necessity.
